### DISCUSSION TITLE
How best to use a record like a dictionary?

### DISCUSSION THREAD

**Discussion by tesujimath** [nushell/nushell#11910] [Q&A]
My understanding is that Nu expects you to use a record for dictionary use cases (aka associative array).

While records are convenient to use when you are accessing known fields, if you are trying to "lookup a key in a dictionary", it seems awkward to me.  Am I missing something here?

Here's a record which has the fullnames for each user in `/etc/passwd`:
```
let name_by_user = (open /etc/passwd | lines | split column ':' user x uid gid name home shell | select user name | transpose -rd)
```

When I want to lookup say `$user` I am doing this:
```
($name_by_user | get -i $user)
```

Updating is similarly awkward:
```
($name_by_user | upsert $user $newname)
```

This seems quite unnatural, compared with e.g. `$name_by_user[$user]`.

Am I just not grokking the Nu way yet?

**Comment by fdncred** [comment]
It depends on what you're doing. I think it's sometimes easier to deal with list of records aka tables using `where` and `==` or `=~`.
```nushell
open /etc/passwd | lines | split column ':' user x uid gid name home shell | where user == root
```
One thing to keep in mind is that `$name_by_user[$user]` is modifying an array mutably whereas `$name_by_user | upsert $user $newname` is not. I'm pretty sure it's just making a new one with the modified items.

**Reply by tesujimath** [reply]
Ah, I had certainly overlooked the mutable/immutable difference here, thanks!

But I'm not sure whether mutable update with that syntax is allowed?

```
> mut s = {a: 1}
> let field = "c"

> $s[$field] = 3
Error: nu::parser::parse_mismatch

  × Parse mismatch during operation.
   ╭─[entry #18:1:1]
 1 │ $s[$field] = 3
   · ─────┬────
   ·      ╰── expected valid variable name
   ╰────

```

**Reply by fdncred** [reply]
It's not allowed like that. However, you could do this.
```nushell
mut s = {a: 1}
$s.a = 2
```

**Reply by tesujimath** [reply]
Yes, that's my original observation that this doesn't really work for "update the entry with this key in my dictionary".  I think `a` has to be a literal, can't be taken from a variable.
